[SPIP] Saved Filters

gqcorneby opened the pull request:
### :pushpin: References
-   **Issue:** Closes #8697huw58

### Important notes
After installing the app, the save filters feature can only be accessed if current user has one of the following roles. Assume each item on the list has the permissions the previous one has:

1. Saved Filters  - allows users to view public filters and their own private filters (if ever they have any)
2. Saved Filters Create - gives user ability to save/edit/delete their own private filters
3. Saved Filters Create Public - gives user ability to save/edit/delete public filters they own
4. Saved Filters Admin  - gives user ability to save/edit/delete public filters regardless of owner

### :memo: Implementation

- Added ability to save filer, select from saved filters, rename, delete, make private/public
- Added option to save as new filter
- added categories in saved filters "Shared Filters" for public and "My Filters" for private
- updated originally proposed model
    - added id for each filter (need this to rename)
    - added path for ou filters (need this to open ou selection dialog)
- namespace `dashboard`, key `savedFilters`  

TODO: Translations.

### :fire: Notes to the reviewer
- for `tSelectSavedFilter` - I added a check `if (!deepEqual(filter.values, appliedFilters))` to prevent "re-applying" the same exact filters if ever they are the same. I was initially worried that it would result to a lot of unnecessary requests to apply the same filter
- not sure if I should add some loading indicator and some kind of notification that an action was done or if there are any errors


### :video_camera: Screenshots/Screen capture
https://github.com/user-attachments/assets/7e2c7976-0f76-47b3-8fa0-78c533e67100

gqcorneby committed:
save filters, rename, make public/private, delete, save as new

eperedo commented on src/modules/util.js:
```diff
@@ -16,6 +16,30 @@ export function arrayToIdMap(array) {
     }, {})
 }
 
+export function deepEqual(obj1, obj2) {
```
We already have Lodash in the repo, but I'm not sure if _.isEqual (see [docs](https://lodash.com/docs/4.17.15#isEqual)) works in the same way. If it doesn't, that's okay—this function works.

eperedo commented on src/actions/savedFilters.js:
```diff
@@ -0,0 +1,111 @@
+    SET_SAVED_FILTERS,
+    sGetActiveFilter,
+    sGetSavedFiltersList,
+} from '../reducers/savedFilters.js'
```
There're some eslint warnings/errors we can fix:

```bash
yarn lint

/dashboard-app-fork/src/reducers/index.js
  17:1   error  `./savedFilters` import should occur before import of `./selected.js`  import/order
  17:26  error  Missing file extension "js" for "./savedFilters"   
```

Only fix warnings/errors related to the code we're introducing in this PR, we can ignore the rest.

eperedo commented on src/pages/view/FilterBar/SaveFilterDialog.js:
```diff
@@ -0,0 +1,89 @@
+                    </Button>
+
+                    <Button key="confirm" destructive onClick={handleConfirm}>
+                        {'Confirm'}
```
Use i18n.t for texts. `i18n.t("Confirm")` and `i18n.t("Cancel")`. Check other components if necessary.

eperedo commented on src/pages/view/FilterBar/SaveFilterDialog.js:
```diff
@@ -0,0 +1,89 @@
+                        selected={visibility}
+                        onChange={handleVisibilityChange}
+                    >
+                        {filterVisibility.map((f) => (
```
Not a big deal, but let's try to avoid one letter variables. `f` -> `filter`, maybe?

eperedo commented on src/pages/view/FilterBar/FilterBar.js:
```diff
@@ -26,16 +57,147 @@ const FilterBar = ({ filters, removeFilter, removeAllFilters }) => {
+        setFilterDialogIsOpen(true)
+        setMoreOptionsIsOpen(false)
+    }, [activeFilter])
+    const handleDeleteFilter = useCallback(async () => {
```
Can we move all of these methods into a hook? Additionally, add spaces between methods to improve readability

eperedo commented on src/pages/view/FilterBar/FilterBar.js:
```diff
@@ -26,16 +57,147 @@ const FilterBar = ({ filters, removeFilter, removeAllFilters }) => {
+    const closeFilterDialog = useCallback(() => {
+        setFilterDialogIsOpen(false)
+        setShowScope(true)
+    }, [activeFilter])
```
I think we don't need `activeFilter` as dependency in this callback, right?

eperedo reviewed (changes requested):
The code looks good. In addition to my comments about the code, I have some UI-related suggestions (though the final decision here is up to @Ramon-Jimenez)

In the rest of the app is typically used the blue one variant for primary actions in buttons.

![image](https://github.com/user-attachments/assets/41e8a8da-7f8a-4c2f-8947-9a646fe650f1)

While saving data to the dataStore is usually fast, for slow connections, it might be necessary to show a loading indicator. Perhaps something [similar to what we implemented here](https://github.com/EyeSeeTea/dashboard-app/pull/1#discussion_r1889364157)

Thanks @gqcorneby! Let me know if you have any questions.

Ramon-Jimenez commented:
Thanks for the review and comments @eperedo . I am aligned with the UI ones, so @gqcorneby  along with the small changes requested in the code by Eduardo, please take care of those ones too.

Also, there are a few GitHubt tests not sucessful, could you please review those and see if we can make them pass properly?

Thanks @gqcorneby !

gqcorneby committed:
address comments.
refactor FilterBar > SavedFilter

gqcorneby committed:
fix lint and test errors

gqcorneby committed:
fix concurrency issue

gqcorneby committed:
saving indicator

eperedo commented on src/pages/view/FilterBar/SaveFilterDialog.js:
```diff
@@ -0,0 +1,95 @@
+                        {i18n.t('Cancel')}
+                    </Button>
+
+                    <Button key="confirm" primary onClick={handleConfirm}>
```
I think we need the loading indicator here as well, right? For slow connections and to prevent user clicking twice.

eperedo commented on src/pages/view/FilterBar/SavedFilterActions.js:
```diff
@@ -0,0 +1,145 @@
+    const [moreOptionsIsOpen, setMoreOptionsIsOpen] = useState(false)
+    const savedFilterHasChanges = useMemo(
+        () => hasActiveSavedFilter && !isEqual(activeFilter.values, filters),
+        [activeFilter, filters]
```
`hasActiveSavedFilter` should be in the dependency array

eperedo commented on src/pages/view/FilterBar/useSaveFilterDialog.js:
```diff
@@ -0,0 +1,43 @@
+        setDialogIsOpen(false)
+        setShowScope(true)
+    }, [])
+    const onConfirm = useCallback(
```
Add some space to keep consistency with the rest of the code

eperedo reviewed (changes requested):
Sorry I wasn't clear enough about where do we need the loading indicator, please check my comments and other minor things. We're very close @gqcorneby!

gqcorneby committed:
add saving indicator in dialog
- loading indicator when fetching dimensions
- add saved filters button width

gqcorneby committed:
Revert name tooltip

gqcorneby committed:
fix css lint error

gqcorneby commented:
Thank you @eperedo! After testing with network throttling, I encountered an error with selecting a saved filter. Applying filters require the dimensions to be loaded so I added a loading state in the list to make sure dimensions are loaded before selection. I also added a width to the "saved filters" button to limit the length for long names

gqcorneby committed:
fix css lint error

eperedo reviewed (approved):
Looks good to me. thanks @gqcorneby!

Ramon-Jimenez reviewed (changes requested):
Thanks @gqcorneby . A few things I noticed after some testing, I think most of them will be pretty easy to fix/adjust:

- [x] If there are no Saved Filters at all, the user doesn't see the option 'Saved filters' at top. I think it should be always there, even though if there are no Saved Filters (in that case we can show something like 'No saved filters available' if you click in the button.

- [x] When a filter is selected, we are changing the 'Saved Filters' text button to use the name of the filter. I think we should keep the text 'Saved Filters' in the button, since the name of the selected filter can be seen in other places. In other words, the button should always read 'Saved Filters' no matter a filter is selected or not.

- [x] When you have a Saved filter selected, and click in the X icon of the filter currently applied, the Saved filter keeps showing as selected in the Saved filter menu, even though the filter is not applied anymore. We should make sure the filter is not selected after this scenario happens.
![image](https://github.com/user-attachments/assets/82f895d0-1936-4a4c-8cb5-6aec99712f4e)
![image](https://github.com/user-attachments/assets/bbaefc34-ab8f-4d0d-9520-e0c43b2e4c25)

- [x] There is an scenario we didn't take into account, but we should support. If there is a Public filter that uses an Org Unit (e.g.: OrgUnit = Albania), and then a user that DO NOT have access to that Org Unit, tries to use that Public filter, currently we are showing like the filter it is applied when, in fact, we are getting a 409 from analytics endpoint because the user doesn't have access to the OrgUnit. We should show an error message in a pop-up saying "The selected Saved Filter cannot be applied because it includes selections that the current user does not have permission to view. Please try a different filter." with an OK button to dismiss the message.

- [x] Good to have: Another scenario that we didn't take into account. Admins can make private their own public filters, but also public filters created by other users. For this last scenario, we should inform the user with a pop-up saying "The public Saved Filter you are attempting to make private was created by <userName>. Do you still want to proceed?" with a Yes and Cancel buttons.

The rest works great, it is going to be very helpful for the users.

Commit:
deactivate selected filter when no filter is applied. Alwasy show button

Commit:
prevent applying filters that contain restricted ou

Commit:
refactor tSelectSavedFilter: reduce nesting

Commit:
clean up

Commit:
show alert when selecting restricted ou

Commit:
prompt when superuser attempts to update a saved filter created by another

Commit:
clean up and remove lint errors

Commit:
fix test

gqcorneby commented:
Hi @Ramon-Jimenez and @eperedo! I addressed the issues from the previous comment and added the visual prompts. 


- **For the 5th item, I  used an existing `ConfirmActionDialog` which looks like this**
![image](https://github.com/user-attachments/assets/89f4b078-c3ce-42ec-a3ed-7d61f91175ed)

- **For the 4th item, I used the dhis2 alert and used a warning flag.** 

![image](https://github.com/user-attachments/assets/ec762645-afbf-4e38-88b6-295d653e0f4b)

eperedo commented on src/actions/savedFilters.js:
```diff
@@ -0,0 +1,136 @@
+import _, { isEqual } from 'lodash'
```
I just realized for lodash they use specific imports instead using the whole library or destructuring probably to reduce the build size.

```ts
import _ from "lodash; ❌ 
import { isEqual } from 'lodash' ❌ 

import isEqual from 'lodash/isEqual.js' ✅ 
import every from 'lodash/every.js'; ✅ 
import some from 'lodash/some.js'; ✅ 
```

Let's follow this convention to keep consistency with the rest of the code.

eperedo commented on src/pages/view/FilterBar/SavedFilterActions.js:
```diff
@@ -0,0 +1,176 @@
+                    onClick={toggleMoreActions}
+                    icon={<IconMore16 color={colors.grey700} />}
+                    component={savedFilterActions}
+                />
```
Children is required for `<DropdownButton />`. can we pass an empty fragment to avoid the error? 

```ts
<DropdownButton ...>
   <></>
</DropdownButton 
```

I know the ideal solution should be to make the prop optional, but let's try to avoid modifying validations for an existing component.

![image](https://github.com/user-attachments/assets/1646e2c5-68b1-4217-8a2e-3876e8259239)

Also add an space between dropdown and `<ConfirmActionDialog`

eperedo commented on src/pages/view/FilterBar/useSaveFilterDialog.js:
```diff
@@ -0,0 +1,54 @@
+    const [showScope, setShowScope] = useState(true)
+
+    const openDialogForNewFilter = useCallback(() => {
+        setFilterDialogData({ id: 'new' })
```
Can we create a constant for the 'new' string? So we can reuse it in both places. 

```ts
const NEW_FILTER_ID = 'new'; // feel free to use another name

...
setFilterDialogData({ id: NEW_FILTER_ID  })

...

const onConfirm = useCallback(
        ({ name, visibility, id }) => {
            doSaveFilter({
                ...activeFilter,
                name,
                visibility,
                id,
                ...(id === NEW_FILTER_ID  && { userName: null, userId: null }),
            }).then((resp) => resp && onClose())
        },
        [activeFilter, onClose, doSaveFilter]
    )
```

Just to be safe (in case we need to use the value elsewhere), we should reuse the variable. This will make refactoring easier if changes are needed in the future.

eperedo reviewed (changes requested):
Thanks @gqcorneby! Let me know if you have any questions related to my comments.

Commit:
cleanup

Commit:
fix lint errors

Ramon-Jimenez commented:
@gqcorneby it is working great! One last addition I think would be useful after testing this, and a couple of quick fixes

- [x] Let's add an icon before the name of each filter in the Saved filters list, for those filters for which the user is the owner. We can include a ToolTip when you hover over the icon saying 'You are the owner of this filter', so it is easy for the user to understand which filters are his/her.

![image](https://github.com/user-attachments/assets/83e532ba-f62d-40f5-80d2-b262588a8dca)

- [x] Let's sort the list of filters, both Shared and Private, alphabetically.

- [x] In the Scope selector, please uppercase the first letter of public and private (Public / Private).

![image](https://github.com/user-attachments/assets/c4028f4f-8e21-442c-bf0f-02e3e99bc916)

- [x] The system allows the user to save a filter with no name, please add a validation so you cannot save a filter unless a name has been entered.

Commit:
require name, owner indicator, scope selector, sort by name

Commit:
lint error

Commit:
prettier

eperedo reviewed (approved):
Looks good, thanks @gqcorneby !

Ramon-Jimenez reviewed (approved):
Thanks @gqcorneby all good now!

Commit:
use position "top" for dialogs

Commit:
use "Saved Filters Admin"

Commit:
search userRoles instead of authorities

Commit:
add unsaved changes indicator
- add hook for saved filter logic in FilterSelector
- add unsaved changes warning

Commit:
add path check
- refactor to use own buildFilter method instead of sGetNamedItemFilters

Commit:
Add warning unsaved changes when renaming/toggling visibility/delete. Combine with not owner message.

gqcorneby commented:
Reference: https://app.clickup.com/t/86988up97 

@Ramon-Jimenez 
Re: Duplicate names
If the name is already in the loaded saved filters list, the error will show up immediately. But, since the saved filters aren't updated real-time when new ones are saved by other users, when we attempting to use a name that already exists but hasn't been fetched yet, the error will only show up after a save attempt. 

https://github.com/user-attachments/assets/6cc5c21e-5b41-4e16-abdb-18eec78676e1

The duplicate name error when attempting to toggle looks like this
![image](https://github.com/user-attachments/assets/5d1685af-24c3-4c89-ad64-17f773ac4785)

Commit:
add duplicate name valication check

Commit:
refactor error handling on toggle visibility

Commit:
fix error messages for translation

Commit:
feat: limit filter actions to admin and create roles

gqcorneby committed:
feat: use alert dialog instead of useAlert to show at top instead of bottom

gqcorneby committed:
fix: remove save option for users without appropriate role

gqcorneby committed:
fix: remove deselecting behavior

gqcorneby committed:
fix: show save button based on role

gqcorneby committed:
feat: create saved filters permissions based on role
- apply visibility updates on actions based on permissions

gqcorneby committed:
fix: add default permissions

gqcorneby committed:
fix: add default permissions

gqcorneby committed:
fix: update default to empty object

gqcorneby committed:
feat: add dashboardId to saved Filters. Only display filters for selected dashboard

gqcorneby committed:
fix: add null check on selectedDashboard to pass tests

gqcorneby committed:
feat: apply dashboardid check in validating name

gqcorneby committed:
fix: pass dashboardId when validating name

gqcorneby committed:
remove logs

gqcorneby committed:
fix: pass selectedDashboard id instead of from activeFilter to handle new filters
